node/actors/bft/src/testonly/ut_harness.rs

```diff
@@ -0,0 +1,353 @@
+                payload: payload.hash(),
+            },
+            proposal_payload: Some(payload),
+            justification: self.rng().gen(),
+        };
+
+        mutate_fn(&mut msg);
+
+        self.consensus
+            .inner
+            .secret_key
+            .sign_msg(ConsensusMsg::LeaderPrepare(msg))
+    }
+
+    pub(crate) fn new_current_replica_commit(
+        &self,
+        mutate_fn: impl FnOnce(&mut ReplicaCommit),
+    ) -> Signed<ConsensusMsg> {
+        let mut msg = ReplicaCommit {
+            protocol_version: validator::CURRENT_VERSION,
+            view: self.consensus.replica.view,
+            proposal: self.consensus.replica.high_qc.message.proposal,
+        };
+
+        mutate_fn(&mut msg);
+
+        self.consensus
+            .inner
+            .secret_key
+            .sign_msg(ConsensusMsg::ReplicaCommit(msg))
+    }
+
+    pub(crate) fn new_rnd_leader_commit(
+        &mut self,
+        mutate_fn: impl FnOnce(&mut LeaderCommit),
+    ) -> Signed<ConsensusMsg> {
+        let mut msg = LeaderCommit {
+            protocol_version: validator::CURRENT_VERSION,
+            justification: self.rng().gen(),
+        };
+
+        mutate_fn(&mut msg);
+
+        self.consensus
+            .inner
+            .secret_key
+            .sign_msg(ConsensusMsg::LeaderCommit(msg))
+    }
+
+    pub(crate) async fn new_procedural_leader_prepare(&mut self) -> Signed<ConsensusMsg> {
+        let replica_prepare = self.new_current_replica_prepare(|_| {});
+        self.dispatch_replica_prepare(replica_prepare.clone())
+            .unwrap();
+        self.recv_signed().await.unwrap()
+    }
+
+    pub(crate) async fn new_procedural_replica_commit(&mut self) -> Signed<ConsensusMsg> {
+        let replica_prepare = self.new_current_replica_prepare(|_| {});
+        self.dispatch_replica_prepare(replica_prepare.clone())
+            .unwrap();
+        let leader_prepare = self.recv_signed().await.unwrap();
+        self.dispatch_leader_prepare(leader_prepare).await.unwrap();
+        self.recv_signed().await.unwrap()
+    }
+
+    pub(crate) async fn new_procedural_leader_commit(&mut self) -> Signed<ConsensusMsg> {
+        let replica_prepare = self.new_current_replica_prepare(|_| {});
+        self.dispatch_replica_prepare(replica_prepare.clone())
+            .unwrap();
+        let leader_prepare = self.recv_signed().await.unwrap();
+        self.dispatch_leader_prepare(leader_prepare).await.unwrap();
+        let replica_commit = self.recv_signed().await.unwrap();
+        self.dispatch_replica_commit(replica_commit).unwrap();
+        self.recv_signed().await.unwrap()
+    }
+
+    #[allow(clippy::result_large_err)]
+    pub(crate) fn dispatch_replica_prepare_many(
+        &mut self,
+        messages: Vec<ReplicaPrepare>,
+        keys: Vec<SecretKey>,
+    ) -> Result<(), ReplicaPrepareError> {
+        messages
+            .iter()
+            .enumerate()
```

Comment:
- `messages` can be consumed via `into_iter()` to avoid cloning later.
- Enumeration can be avoided by `zip`ping `messages` and `keys`; this is considered more idiomatic.
- AFIACT, `keys` are not consumed by the method, so you can pass `&[SecretKey]` as an arg.

Reply:
> - `messages` can be consumed via `into_iter()` to avoid cloning later.

👍
Fixed.
> - Enumeration can be avoided by `zip`ping `messages` and `keys`; this is considered more idiomatic.

👍
Fixed.

> - AFIACT, `keys` are not consumed by the method, so you can pass `&[SecretKey]` as an arg.

I refrained from using a reference because it overlaps with mutable borrowing in `util.dispatch_replica_prepare_many`. I didn't find a reasonable way to overcome this:

`ut_harness.rs`
```rust
pub(crate) fn keys(&self) -> Vec<&SecretKey> {
    self.keys.iter().collect()
}
``` 
`tests.rs`
```rust
let keys = util.keys();
let res = util.dispatch_replica_prepare_many(
    vec![replica_prepare; consensus_threshold(num_validators)],
    keys,
);
```
```rust
error[E0502]: cannot borrow `util` as mutable because it is also borrowed as immutable
   --> actors/bft/src/replica/tests.rs:364:15
    |
363 |       let keys = util.keys();
    |                  ---- immutable borrow occurs here
364 |       let res = util.dispatch_replica_prepare_many(
    |                 ^    ----------------------------- immutable borrow later used by call
    |  _______________|
    | |
365 | |         vec![replica_prepare; consensus_threshold(num_validators)],
366 | |         keys,
367 | |     );
    | |_____^ mutable borrow occurs here
```